presto-native-execution/src/test/java/com/facebook/presto/nativeworker/TestPrestoNativeJoinQueries.java

```diff
@@ -13,21 +13,95 @@
+import org.testng.annotations.Test;
+
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.any;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.anyTree;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.node;
 
 public class TestPrestoNativeJoinQueries
         extends AbstractTestNativeJoinQueries
 {
     @Override
     protected QueryRunner createQueryRunner() throws Exception
     {
-        return PrestoNativeQueryRunnerUtils.createNativeQueryRunner(true);
+        QueryRunner queryRunner = PrestoNativeQueryRunnerUtils.createNativeQueryRunner(true);
+        queryRunner.installPlugin(new Plugin()
+        {
+            @Override
+            public Iterable<HistoryBasedPlanStatisticsProvider> getHistoryBasedPlanStatisticsProviders()
+            {
+                return ImmutableList.of(new InMemoryHistoryBasedPlanStatisticsProvider());
+            }
+        });
+        return queryRunner;
     }
 
     @Override
     protected ExpectedQueryRunner createExpectedQueryRunner() throws Exception
     {
         return PrestoNativeQueryRunnerUtils.createJavaQueryRunner();
     }
+
+    @Test
+    public void testJoinNull()
+    {
+        String sql = "SELECT\n" +
+                "    O.totalprice,\n" +
+                "    C.name\n" +
+                "FROM (\n" +
+                "    SELECT\n" +
+                "        totalprice,\n" +
+                "        custkey,\n" +
+                "        orderdate\n" +
+                "    FROM orders\n" +
+                "\n" +
+                "    UNION ALL\n" +
+                "\n" +
+                "    SELECT\n" +
+                "        *\n" +
+                "    FROM (\n" +
+                "        VALUES\n" +
+                "            (10.0, NULL, '1995-09-20')\n" +
+                "    )\n" +
+                ") O\n" +
+                "JOIN (\n" +
+                "    SELECT\n" +
+                "        name,\n" +
+                "        custkey\n" +
+                "    FROM customer\n" +
+                "\n" +
+                "    UNION ALL\n" +
+                "\n" +
+                "    SELECT\n" +
+                "        *\n" +
+                "    FROM (\n" +
+                "        VALUES\n" +
+                "            ('unknown', NULL)\n" +
+                "    ) t(name, custkey)\n" +
+                ") C\n" +
+                "    ON C.custkey = O.custkey\n" +
+                "    AND O.orderdate like '1995%'";
+        assertPlan(sql, anyTree(node(JoinNode.class, anyTree(any()), anyTree(any())).withOutputRowCount(Double.NaN).withJoinStatistics(Double.NaN, Double.NaN, Double.NaN, Double.NaN)));
+
+        getQueryRunner().execute(sql);
+        getHistoryProvider().waitProcessQueryEvents();
+        assertPlan(sql, anyTree(node(JoinNode.class, anyTree(any()), anyTree(any())).withOutputRowCount(2204).withJoinStatistics(1501, 1, 2205, 1)));
```

Comment:
Here, we verify that HBO was able to apply the stats from previous run to a new run?

Would it make sense to also verify that we can run the query and get correct results?

Is it possible to verify that HBO applied skew mitigation to the query?

Reply:
Added verification check. HBO applied skew mitigation check seems hard to do for now. Looks like we have hardcord null element thresholds(100K) - we'll need bigger tables in test for that. So I skipped that for now, and added a comment

https://github.com/prestodb/presto/blob/eb94b3f4b83dae444519c618b42b4c853d4380b0/presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java#L196

Reply:
> Looks like we have hardcord null element thresholds(100K) 

Got it. @feilong-liu @mlyublena  It would be nice to document this.